pkg/core/blockchain.go

```diff
@@ -1430,15 +1430,33 @@ func ScriptFromWitness(hash util.Uint160, witness *transaction.Witness) ([]byte,
 
 // Various witness verification errors.
 var (
-	ErrWitnessHashMismatch = errors.New("witness hash mismatch")
-	ErrInvalidSignature    = errors.New("signature check failed")
+	ErrWitnessHashMismatch         = errors.New("witness hash mismatch")
+	ErrInvalidSignature            = errors.New("signature check failed")
+	ErrUnknownVerificationContract = errors.New("unknown verification contract")
+	ErrInvalidVerificationContract = errors.New("verification contract is missing `verify` method")
 )
 
 // verifyHashAgainstScript verifies given hash against the given witness.
 func (bc *Blockchain) verifyHashAgainstScript(hash util.Uint160, witness *transaction.Witness, interopCtx *interop.Context, useKeys bool, gas int64) error {
-	verification, err := ScriptFromWitness(hash, witness)
-	if err != nil {
-		return err
+	var offset int
+	var initMD *manifest.Method
+	verification := witness.VerificationScript
+	if len(verification) != 0 {
+		if witness.ScriptHash() != hash {
+			return ErrWitnessHashMismatch
+		}
+	} else {
+		cs, err := interopCtx.DAO.GetContractState(hash)
```

Comment:
It's a separate non-preview3-compatible feature, it shouldn't be a part of this PR.

Reply:
Are you sure? It was 4 months ago neo-project/neo#1481

Reply:
Let me check, I was thinking more of neo-project/neo#1800.

Reply:
Yeah, this part is really preview3, but it should substitute `ScriptFromWitness()` then.

Reply:
`ScriptFromWitness` has not necessary context, and its signature will become no so understandable.
I am thinking about making `verifyHashAgainstScript` public.

Reply:
That's a good option too, simplifying consensus a bit along the way.

Reply:
I'd postpone it until after release, as it requires some refactoring of both `core` and `consensus` to do right.